### ISSUE TITLE
Bug: `exactOptionalPropertyTypes` causes type errors when using plugins that were not built with this option

### ISSUE THREAD

**Issue by sebastian-altamirano** [eslint/markdown#402]
### Environment

Node version: v22.15.0
npm version: v11.3.0
Local ESLint version: v9.27.0 (Currently used)
Global ESLint version: Not found
Operating System: linux 5.15.167.4-microsoft-standard-WSL2

### What parser are you using?

Default (Espree)

### What did you do?

I have installed some plugins and configured ESLint to use them. ESLint works fine, but TypeScript throws some type errors related to `exactOptionalPropertyTypes`, which I have set to `true` in my `tsconfig.json` since I'm extending from `@tsconfig/strictest`.

### What did you expect to happen?

There should be no type errors when using `exactOptionalPropertyTypes`.

### What actually happened?

Type errors occurred.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/vitejs-vite-gantpukn?file=eslint.config.mjs,tsconfig.json&view=editor

### Participation

- [x] I am willing to submit a pull request for this issue.

### Additional comments

I have 2 reproductions:

- https://stackblitz.com/edit/vitejs-vite-gantpukn?file=eslint.config.mjs,tsconfig.json&view=editor: Contains a minimal reproducible example. It reproduces the error using a single plugin `@eslint/markdown`.
- https://stackblitz.com/edit/stackblitz-starters-rh2kntpu?file=eslint.config.mjs,tsconfig.config.json&view=editor: This is the configuration where I experienced the issue. It makes use of many plugins, so there are more type errors. You may notice that `eslint` does not work in this repro. This is because npm fails to install `eslint-import-resolver-typescript`, which requires native modules. It works if you download the code and run it locally. You can also make it work by force installing `@unrs/resolver-binding-wasm32-wasi` as suggested in https://github.com/import-js/eslint-import-resolver-typescript/issues/431. I did not do this because it breaks `npm i` and I did not want to break the repo.

According to @nzakas: "The `Plugin` type exported from eslint is the thing that needs updating to address this." (https://github.com/eslint/markdown/pull/391#issuecomment-2912790683)

**Comment by sebastian-altamirano** [comment]
What if `exactOptionalPropertyTypes` is added only to `tests/types/tsconfig.json`? This should help detect future incompatibilities. [The tests already use a stricter configuration than the code](https://github.com/eslint/markdown/blob/v6.4.0/tests/types/tsconfig.json#L7).

**Comment by fasttime** [comment]
> [@fasttime](https://github.com/fasttime) can you summarize? I can't quite follow the thread here.

Understandable, since the original issue includes multiple problems. The specific problem concerning this repo is that the generated types aren't usable with ESLint when a user has the [`exactOptionalPropertyTypes`](https://www.typescriptlang.org/tsconfig/#exactOptionalPropertyTypes) option enabled in its TypeScript config. When this option is set, TypeScript does not allow assigning `undefined` to an optional property whose type does not include `undefined`. This is problematic for `eslint/markdown` because the autogenerated types for the `processor` config (`plugin.configs.processor`) contain some properties that are being inconsistently typed as `undefined`, for example `files`:

```ts
            files?: undefined;
```

(from https://app.unpkg.com/@eslint/markdown@6.4.0/files/dist/esm/index.d.ts#L187)

ESLint will throw an error if a config object has `files` explicitly set to `undefined`,  and the types for `Linter.Config["files"]` reflect this fact by not including `undefined` explicitly:

```ts
		files?: Array<string | string[]>;
```

(from https://github.com/eslint/eslint/blob/v9.28.0/lib/types/index.d.ts#L1778)

The result is that `plugin.configs.processor` is not assignable to `Linter.Config[]` when `exactOptionalPropertyTypes` is on and ultimately, the plugin object is not assignable to `ESLint.Plugin`.

This isn't difficult to fix, but there are a few ways to do so, so I'd like to know what others think is the best solution.

**Comment by nzakas** [comment]
Thanks, that's very helpful. So my understanding is that `tsc` is generating a type for `processor` config that doesn't line up with the `Config` definition in the `eslint` package?

If that's so, couldn't we just type the export of `index.js` as `ESLint.Plugin`, like this?
https://github.com/eslint/markdown/pull/408

**Comment by nzakas** [comment]
I'd much prefer a solution that doesn't put the responsibility for this on plugin authors, and if it must be on them, then to be as light-touch as possible.

Is there a reason we can't redefined the `files` property in ESLint to allow `undefined`? Wouldn't that solve it without needing plugins to make changes?

**Comment by nzakas** [comment]
It seems like the properties of `LegacyConfig` already follow this pattern of optional properties also being set to undefined:
https://github.com/eslint/eslint/blob/main/lib/types/index.d.ts#L1453

> TS users will be no longer notified by the compiler if they accidentally set files to a potentially undefined value in their configs, which could happen because of a mistake in the logic. 

How important is this?

@mdjermanovic what do you think?

**Comment by mdjermanovic** [comment]
> > TS users will be no longer notified by the compiler if they accidentally set files to a potentially undefined value in their configs, which could happen because of a mistake in the logic.
> 
> How important is this?
> 
> [@mdjermanovic](https://github.com/mdjermanovic) what do you think?

If we don't intend to allow `undefined` values in runtime, then I think it is important that the types also don't allow them, so that users are notified when attempting to set a value that is potentially `undefined`, as that's what types are for.

**Comment by nzakas** [comment]
This is where I'm not quite sure what is idiomatic for TypeScript. If a property is optional, then setting it to undefined seems like the same thing to me. As already noted, this is how `LegacyConfig` already works. 

In any case, I'd like a solution that doesn't require every plugin to do strange things just to get its types to validate. I don't find any of the solutions listed in https://github.com/eslint/markdown/issues/402#issuecomment-2930878167 satisfying.

Is this something we can solve with [`definePlugin`](https://github.com/eslint/rfcs/pull/132)?

cc @JoshuaKGoldberg

**Comment by JoshuaKGoldberg** [comment]
> If a property is optional, then setting it to undefined seems like the same thing to me.

The difference being picked up on by `exactOptionalPropertyTypes` is whether:

* The object does not have the key at all
* The object has the key, and the corresponding value is `undefined`

That difference doesn't matter often in real-world code. Hence it not being on by default in TypeScript. But making a distinction is more technically correct. 

+1 to what lishaduck noted: even though a lot of existing projects don't use `exactOptionalPropertyTypes`, it's going to get more and more popular over time. And a nontrivial % of projects use it already. My general opinion is that it'd be best to change the exported types in `esm/index.d.ts` to be more precise, so as to fix type errors for users who enable `exactOptionalPropertyTypes`.

> Build type declarations with `exactOptionalPropertyTypes`

I think this would be the optimal strategy. I'm not sure I follow the full comment though @fasttime, is there any reason not to want to turn `exactOptionalPropertyTypes` on always? 

Note that practically, it's blocked on also enabling `strictNullChecks`. Which causes 79 type errors when enabled. Most of them are from `@types/unist` indicating node properties as undefined. Which is a separate issue: the types should mark a difference between _"standard node with position"_ vs. _"generated node that doesn't have a position"_.

Related conversation about enabling some of the strict flags in ESLint code: https://github.com/eslint/rewrite/issues/118

Failing enabling new compiler options...

> Type config explicitly

This seems like a reasonable good next step to me. Confirmed it fixes & simplifies the generated `esm/index.d.ts` for me locally.

**Comment by fasttime** [comment]
> is there any reason not to want to turn `exactOptionalPropertyTypes` on always?
> 
> Note that practically, it's blocked on also enabling `strictNullChecks`. Which causes 79 type errors when enabled. Most of them are from `@types/unist` indicating node properties as undefined.

This is pretty much the main reason why `exactOptionalPropertyTypes` can't be enabled by default, I think. Enabling `strictNullChecks` would require changes in multiple repos. That would fix the bug described in this issue, but it seems to me that there are simplex methods to accomplish this particular fix.

**Comment by nzakas** [comment]
@JoshuaKGoldberg do you think that the `definePlugin()` function could help with this problem?

**Comment by JoshuaKGoldberg** [comment]
I do think so, yes. Or at least it _should_. I'm glad we're having this discussion now; IMO this is a good note -> test case to add to https://github.com/eslint/rfcs/pull/132.

**Comment by nzakas** [comment]
Then I'd prefer to hold off until we finalize `definePlugin()` to see if that can address this issue holistically.

As mentioned previously, I really don't like forcing every plugin developer to jump through hoops to get the types correct. This seems like something that should "just work" out of the box in some way.

**Comment by JoshuaKGoldberg** [comment]
I added a _Plugin Testing_ section to https://github.com/eslint/rfcs/pull/132 noting this issue. https://github.com/eslint/rfcs/pull/132/commits/6f385015ff0664ff4c607ba66c69aecad813d8bb

**Comment by lishaduck** [comment]
Not stale
